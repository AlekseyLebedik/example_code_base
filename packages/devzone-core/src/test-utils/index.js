/* eslint-disable import/no-extraneous-dependencies */
import { shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

import rootReducer, { INITIAL_STATE } from '../modules/reducers';

configure({ adapter: new Adapter() });
/*
 * Repeatedly render a component tree using enzyme.shallow() until
 * finding and rendering TargetComponent.
 *
 * This is useful for testing a component wrapped in one or more
 * HOCs (higher order components).
 *
 * The `componentInstance` parameter is a React component instance.
 * Example: <MyComponent {...props} />
 *
 * The `TargetComponent` parameter is the React class (or function) that
 * you want to retrieve from the component tree.
 *
 * Copied from: https://github.com/mozilla/addons-frontend
 */
export function shallowUntilTarget(
  componentInstance,
  TargetComponent,
  { maxTries = 10, shallowOptions, _shallow = shallow, rendered = false } = {}
) {
  if (!componentInstance) {
    throw new Error('componentInstance parameter is required');
  }
  if (!TargetComponent) {
    throw new Error('TargetComponent parameter is required');
  }

  let root = rendered
    ? componentInstance
    : _shallow(componentInstance, shallowOptions);

  if (typeof root.type() === 'string') {
    // If type() is a string then it's a DOM Node.
    // If it were wrapped, it would be a React component.
    throw new Error('Cannot unwrap this component because it is not wrapped');
  }

  for (let tries = 1; tries <= maxTries; tries += 1) {
    if (root.is(TargetComponent)) {
      // Now that we found the target component, render it.
      return root.shallow(shallowOptions);
    }
    // Unwrap the next component in the hierarchy.
    root = root.dive();
  }

  throw new Error(
    `Could not find ${TargetComponent} in rendered instance: ${componentInstance};
    gave up after ${maxTries} tries`
  );
}

export const createTestStore = (
  reducer = rootReducer,
  initialState = INITIAL_STATE
) => {
  const store = {
    state: reducer(initialState, {}),
  };
  const getState = () => store.state;
  const dispatch = action => {
    if (typeof a === 'function') {
      action(dispatch, getState);
    } else {
      store.state = reducer(store.state, action);
    }
  };
  store.dispatch = dispatch;
  return store;
};
