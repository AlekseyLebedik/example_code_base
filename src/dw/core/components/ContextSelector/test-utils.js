import {
  ServiceEndpoints,
  ServiceKeyMapping,
} from 'dw/online-configuration/constants';
import { CONTEXT_TYPE_TITLE, CONTEXT_TYPE_PLATFORM } from './constants';

export const DEFAULT_TITLE_CONTEXT = '5577';
export const PLATFORM_CONTEXTS = ['5577-ps4', '5577-xb1'];
export const DEFAULT_PLATFORM_CONTEXT = PLATFORM_CONTEXTS[0];

export const mockState = ({
  serviceNames = [],
  titleContext = DEFAULT_TITLE_CONTEXT,
  platformContexts = PLATFORM_CONTEXTS,
  usesMulticontext = true,
  platformEndpoints = [],
  userId,
}) => {
  const selectableContexts = Array.isArray(platformContexts)
    ? platformContexts
    : [platformContexts];
  const services = Array.isArray(serviceNames) ? serviceNames : [serviceNames];
  const usePlatform = Array.isArray(platformEndpoints)
    ? platformEndpoints
    : [platformEndpoints];
  const state = {
    user: {
      projects: [
        {
          titles: [
            {
              id: 5577,
              environments: [
                { shortType: 'dev', options: { is_multicontext: true } },
              ],
            },
          ],
        },
      ],
    },
    Components: {
      App: {
        currentTitleEnv: {
          usesMulticontext,
        },
      },
      ContextSelector: {
        Available: services.reduce(
          (acc, serviceName) => ({
            ...acc,
            [serviceName]: {
              data: [
                ...selectableContexts.map((context, idx) => ({
                  id: idx + 1,
                  name: context,
                  userSelectable: true,
                  type: CONTEXT_TYPE_PLATFORM,
                })),
                {
                  id: selectableContexts.length + 1,
                  name: titleContext,
                  userSelectable: false,
                  type: CONTEXT_TYPE_TITLE,
                },
              ],
            },
          }),
          {}
        ),
        userId,
        Registry: services.reduce(
          (acc, serviceName) => ({
            ...acc,
            [serviceName]: {
              data: Object.values(
                ServiceEndpoints[ServiceKeyMapping[serviceName]]
              ).map(endpoint => ({
                endpoint,
                type: usePlatform.includes(endpoint)
                  ? CONTEXT_TYPE_PLATFORM
                  : CONTEXT_TYPE_TITLE,
              })),
            },
          }),
          {}
        ),
      },
      TitleSelector: {
        currentTitle: { id: 5577 },
        currentEnv: { shortType: 'dev' },
      },
    },
  };
  return state;
};
