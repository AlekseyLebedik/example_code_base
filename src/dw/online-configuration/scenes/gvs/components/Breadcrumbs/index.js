import React, { useCallback, useMemo } from 'react';
import PropTypes from 'prop-types';
import cn from 'classnames';
import { useSelector } from 'react-redux';
import { useHistory, useLocation, useParams } from 'react-router-dom';

import { makeStyles } from '@material-ui/core/styles';
import Typography from '@material-ui/core/Typography';
import Breadcrumbs from '@material-ui/core/Breadcrumbs';

import { useEnvScopeUri, useScopes } from '../../graphql/hooks';
import CreateScopeButton from '../CreateScopeButton';
import GameBuildSelector from '../GameBuildSelector';

const useStyles = makeStyles(theme => ({
  container: {
    marginTop: theme.spacing(1),
    marginBottom: theme.spacing(1),
    marginLeft: ({ scene }) => (scene === 'groups' ? theme.spacing(2) : 0),
    position: 'relative',
    display: 'flex',
    justifyContent: 'space-between',
    alignItems: 'center',
  },
  breadcrumb: {
    display: 'flex',
  },
  separator: {
    margin: `0 ${theme.spacing(2)}px`,
  },
  buildSelectRoot: {
    marginBottom: 0,
    marginTop: '1px',
  },
  buildSelect: {
    padding: 0,
    paddingLeft: '3px',
    minWidth: '150px',
    ...theme.typography.subtitle2,
  },
  disabledBreadcrumb: {
    color: 'rgba(0, 0, 0, 0.5)',
  },
  active: {
    fontWeight: '600',
  },
}));

const GVSBreadcrumbs = React.forwardRef(
  ({ scopeOverride, hideGameBuild, variant }, ref) => {
    const { scopeURI: initialScopeURI, titleId, env, ...params } = useParams();
    const classes = useStyles(params);
    const titleName = useSelector(
      state => state.Components.TitleSelector.currentTitle?.name
    );
    const projectName = useSelector(
      state => state.Components.TitleSelector.currentProject?.name
    );
    const scopeURI = useMemo(
      () => scopeOverride || initialScopeURI,
      [scopeOverride, initialScopeURI]
    );
    const location = useLocation();
    const history = useHistory();
    const { refetch } = useScopes();
    const { envScopeUri } = useEnvScopeUri();
    const [franchise, game] = useMemo(
      () => (envScopeUri ? envScopeUri.split(':') : []),
      [envScopeUri]
    );
    const onScopeChange = useCallback(
      newScopeUri => {
        history.push({
          pathname: location.pathname.replace(scopeURI, newScopeUri),
          search: location.search,
        });
      },
      [history, location]
    );
    return (
      <div className={classes.container} ref={ref}>
        <div className={classes.breadcrumb}>
          <Breadcrumbs
            separator="â€º"
            aria-label="breadcrumb"
            classes={{ separator: classes.separator }}
          >
            {variant === 'full' && [
              <Typography
                key="env"
                variant="subtitle2"
                className={classes.disabledBreadcrumb}
              >
                {env.toUpperCase()}
              </Typography>,
              <Typography
                key="franchise"
                variant="subtitle2"
                className={cn({
                  [classes.disabledBreadcrumb]: scopeURI !== franchise,
                  [classes.active]: scopeURI === franchise,
                })}
              >
                {franchise.toUpperCase()}
              </Typography>,
              <Typography
                key="game"
                variant="subtitle2"
                className={cn({
                  [classes.disabledBreadcrumb]:
                    scopeURI !== [franchise, game].join(':'),
                  [classes.active]: scopeURI === [franchise, game].join(':'),
                })}
              >
                {projectName || game.toUpperCase()}
              </Typography>,
            ]}
            <Typography
              key="title"
              variant="subtitle2"
              color="textPrimary"
              className={cn({
                [classes.disabledBreadcrumb]: scopeURI !== envScopeUri,
                [classes.active]: scopeURI === envScopeUri,
              })}
            >
              ({titleId}) - {titleName}
            </Typography>
            {!hideGameBuild && (
              <GameBuildSelector
                scopeURI={scopeURI}
                scopeRoot={envScopeUri}
                onScopeChange={onScopeChange}
                disabled={scopeURI === envScopeUri}
              />
            )}
          </Breadcrumbs>
          <CreateScopeButton
            refetchScopes={refetch}
            onScopeChange={onScopeChange}
          />
        </div>
      </div>
    );
  }
);
GVSBreadcrumbs.propTypes = {
  scopeOverride: PropTypes.string,
  hideGameBuild: PropTypes.bool,
  variant: PropTypes.string,
};
GVSBreadcrumbs.defaultProps = {
  scopeOverride: undefined,
  hideGameBuild: false,
  variant: 'full',
};

export default GVSBreadcrumbs;
