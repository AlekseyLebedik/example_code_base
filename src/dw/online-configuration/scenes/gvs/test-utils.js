/* eslint-disable import/no-extraneous-dependencies */
import React from 'react';
import { shallow } from 'enzyme';
import { MockedProvider } from '@apollo/client/testing';
import createStore from 'dw/online-configuration/store';

import { mocks as baseMocks } from './graphql/mocks';

jest.mock('react-redux', () => {
  return {
    ...jest.requireActual('react-redux'),
    useDispatch: jest.fn(),
    useSelector: jest.fn(),
  };
});

const {
  useSelector: mockUseSelector,
  useDispatch: mockUseDispatch,
} = require('react-redux');

const setupStore = ({ scopeURI }) => {
  const { store: mockStore } = createStore();
  mockUseDispatch.mockImplementation(
    () => action => mockStore.dispatch(action)
  );
  mockStore.dispatch({
    type: 'APP_FETCH_TITLE_ENVIRONMENT_SUCCESS',
    env: {
      options: {
        scopeURI,
      },
    },
  });
  mockStore.dispatch({
    type: 'TITLE_SELECTOR.CHANGE_TITLE',
    title: { name: 'BO5 CrossPlay Scratch Dev' },
    project: { name: 'Call of Duty: Black Ops 5' },
  });
  mockUseSelector.mockImplementation(selector =>
    selector(mockStore.getState())
  );
};

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useParams: jest.fn(),
  useLocation: jest.fn(),
  useRouteMatch: jest.fn(),
  useHistory() {
    return {
      push: jest.fn(),
    };
  },
}));

const {
  useParams: mockUseParams,
  useLocation: mockUseLocation,
  useRouteMatch: mockUseRouteMatch,
} = require('react-router-dom');

export const setupUseParams = (props, search) => {
  mockUseParams.mockReturnValue({
    ...props,
  });
  mockUseLocation.mockReturnValue({
    search,
  });
  mockUseRouteMatch.mockReturnValue({
    params: props,
  });
};

jest.mock('@apollo/client', () => ({
  ...jest.requireActual('@apollo/client'),
  useQuery: jest.fn(),
  useMutation: jest.fn(),
  useApolloClient: jest.fn(),
}));
const { useQuery, useMutation, useApolloClient } = require('@apollo/client');

jest.mock('@demonware/devzone-core/modules/user/hooks', () => ({
  useUserProfileActions() {
    return { user: { profile: { id: 1, name: 'testuser' } } };
  },
}));

jest.mock('dw/core/components/TitleEnvSelect/hooks');

const { useEnvironments } = require('dw/core/components/TitleEnvSelect/hooks');

const mockError = jest.fn();
const mockSuccess = jest.fn();
jest.mock('dw/core/hooks', () => ({
  ...jest.requireActual('dw/core/hooks'),
  useCurrentEnvPermission: jest.fn(),
  useSnackbar: jest.fn(),
}));

const { useCurrentEnvPermission, useSnackbar } = require('dw/core/hooks');

export const renderWithApolloClient = (
  Component,
  { params, scopeURI, search, extraMocks = [], permission = true, snackbar }
) => {
  useEnvironments.mockReturnValue({
    loading: false,
    environments: [
      {
        title: { id: String(params?.titleId || 1) },
        environment: { shortType: params?.env || 'dev', options: { scopeURI } },
      },
    ],
  });
  useCurrentEnvPermission.mockReturnValue(permission);
  useSnackbar.mockReturnValue({
    error: snackbar?.error || mockError,
    success: snackbar?.success || mockSuccess,
  });
  setupStore({ scopeURI });
  setupUseParams(params, search);
  const mocks = [...baseMocks, ...extraMocks];
  const wrapper = shallow(
    <MockedProvider mocks={mocks} addTypename={false}>
      {Component}
    </MockedProvider>
  );

  const { client } = wrapper.props();
  const { useQuery: useOrigQuery, useMutation: useOrigMutation } =
    jest.requireActual('@apollo/client');
  useQuery.mockImplementation((q, options) => {
    const result = useOrigQuery(q, {
      ...options,
      client,
      fetchPolicy: 'cache-first',
    });
    return { ...result, refetch: jest.fn() };
  });
  useMutation.mockImplementation((m, options) =>
    useOrigMutation(m, { ...options, client })
  );
  useApolloClient.mockReturnValue(client);
  return wrapper.shallow().shallow().shallow().shallow();
};
