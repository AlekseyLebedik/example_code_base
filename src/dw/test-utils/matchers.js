expect.extend({
  toAsyncDispatch(received, ...expectations) {
    const dispatchMock = jest.fn();
    const dispatch = action =>
      typeof action === 'function'
        ? action(act => {
            dispatchMock(act);
            return act;
          }, received.getState)
        : dispatchMock(action);
    received(dispatch);
    return expectations
      .map(expected => {
        // Calls a matcher for earch expectation argument
        try {
          // eslint-disable-next-line jest/no-standalone-expect
          expect(dispatchMock).toHaveBeenCalledWith(
            expect.objectContaining(expected)
          );
          return {
            pass: true,
            message: () => `${received} dispatches ${JSON.stringify(expected)}`,
          };
        } catch (err) {
          const message = () =>
            err.matcherResult && err.matcherResult.message
              ? err.matcherResult.message()
              : err.toString();
          return { pass: false, message };
        }
      })
      .reduce(
        // reduces to a single result
        (acc, cur) => ({
          pass: this.isNot ? cur.pass || acc.pass : cur.pass && acc.pass,
          message: () => `${cur.message()}\n${acc.message()}`,
        }),
        { pass: !this.isNot, message: () => '' }
      );
  },
});
