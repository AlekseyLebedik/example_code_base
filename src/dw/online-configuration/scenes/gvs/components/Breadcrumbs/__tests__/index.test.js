import React from 'react';

import wait from 'dw/test-utils/wait';
import { renderWithApolloClient } from '../../../test-utils';

import Breadcrumbs from '../index';

const envScopeURI = 'cod:iw8:5830';
const buildScopeURI = 'cod:iw8:5830:tu_1';

jest.mock('dw/core/components/Select', () => 'Select');

describe('GVS Breadcrumbs', () => {
  it('render title as active', async () => {
    const wrapper = renderWithApolloClient(<Breadcrumbs />, {
      params: {
        env: 'dev',
        titleId: '5830',
        scopeURI: envScopeURI,
      },
      scopeURI: envScopeURI,
    });
    await wait(0);
    expect(wrapper).toMatchSnapshot();
  });
  it('render game build as active', async () => {
    const wrapper = renderWithApolloClient(<Breadcrumbs />, {
      params: {
        env: 'dev',
        titleId: '5830',
        scopeURI: buildScopeURI,
      },
      scopeURI: envScopeURI,
    });
    await wait(0);
    expect(wrapper).toMatchSnapshot();
  });
});
