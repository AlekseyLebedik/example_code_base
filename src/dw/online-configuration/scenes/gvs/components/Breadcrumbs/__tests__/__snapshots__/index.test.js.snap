// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GVS Breadcrumbs render game build as active 1`] = `
<div
  className="makeStyles-container-1 makeStyles-container-9"
>
  <div
    className="makeStyles-breadcrumb-2"
  >
    <WithStyles(ForwardRef(Breadcrumbs))
      aria-label="breadcrumb"
      classes={
        Object {
          "separator": "makeStyles-separator-3",
        }
      }
      separator="›"
    >
      <WithStyles(ForwardRef(Typography))
        className="makeStyles-disabledBreadcrumb-6"
        key="env"
        variant="subtitle2"
      >
        DEV
      </WithStyles(ForwardRef(Typography))>
      <WithStyles(ForwardRef(Typography))
        className="makeStyles-disabledBreadcrumb-6"
        key="franchise"
        variant="subtitle2"
      >
        COD
      </WithStyles(ForwardRef(Typography))>
      <WithStyles(ForwardRef(Typography))
        className="makeStyles-disabledBreadcrumb-6"
        key="game"
        variant="subtitle2"
      >
        Call of Duty: Black Ops 5
      </WithStyles(ForwardRef(Typography))>
      <WithStyles(ForwardRef(Typography))
        className="makeStyles-disabledBreadcrumb-6"
        color="textPrimary"
        key="title"
        variant="subtitle2"
      >
        (
        5830
        ) - 
        BO5 CrossPlay Scratch Dev
      </WithStyles(ForwardRef(Typography))>
      <GameBuildSelector
        disabled={false}
        onScopeChange={[Function]}
        scopeRoot="cod:iw8:5830"
        scopeURI="cod:iw8:5830:tu_1"
      />
    </WithStyles(ForwardRef(Breadcrumbs))>
    <CreateScopeButton
      onScopeChange={[Function]}
      refetchScopes={[MockFunction]}
    />
  </div>
</div>
`;

exports[`GVS Breadcrumbs render title as active 1`] = `
<div
  className="makeStyles-container-1 makeStyles-container-8"
>
  <div
    className="makeStyles-breadcrumb-2"
  >
    <WithStyles(ForwardRef(Breadcrumbs))
      aria-label="breadcrumb"
      classes={
        Object {
          "separator": "makeStyles-separator-3",
        }
      }
      separator="›"
    >
      <WithStyles(ForwardRef(Typography))
        className="makeStyles-disabledBreadcrumb-6"
        key="env"
        variant="subtitle2"
      >
        DEV
      </WithStyles(ForwardRef(Typography))>
      <WithStyles(ForwardRef(Typography))
        className="makeStyles-disabledBreadcrumb-6"
        key="franchise"
        variant="subtitle2"
      >
        COD
      </WithStyles(ForwardRef(Typography))>
      <WithStyles(ForwardRef(Typography))
        className="makeStyles-disabledBreadcrumb-6"
        key="game"
        variant="subtitle2"
      >
        Call of Duty: Black Ops 5
      </WithStyles(ForwardRef(Typography))>
      <WithStyles(ForwardRef(Typography))
        className="makeStyles-active-7"
        color="textPrimary"
        key="title"
        variant="subtitle2"
      >
        (
        5830
        ) - 
        BO5 CrossPlay Scratch Dev
      </WithStyles(ForwardRef(Typography))>
      <GameBuildSelector
        disabled={true}
        onScopeChange={[Function]}
        scopeRoot="cod:iw8:5830"
        scopeURI="cod:iw8:5830"
      />
    </WithStyles(ForwardRef(Breadcrumbs))>
    <CreateScopeButton
      onScopeChange={[Function]}
      refetchScopes={[MockFunction]}
    />
  </div>
</div>
`;
